lib/puppet/file_serving/metadata.rb

```diff
@@ -88,6 +90,21 @@ def collect_stat(path)
   def collect(source_permissions = nil)
     real_path = full_path
 
+    if static_catalog
+      if environment && environment.static_catalogs?
+        env_path = File.join(Puppet[:environmentpath], environment.name.to_s) # environment.name is a symbol sometimes
+        puts "Checking if #{real_path} is in #{env_path}"
+        if real_path.start_with? env_path
```

Comment:
This code gets called on both the agent and server (through their respective `:file_*` termini), which is a bit surprising. Also it forces us to thread the environment and static catalog settings through to metadata and create a metadata object that we throw away if it's not inlineable. Also, soon we'll need to inline ["search" metadata](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file.rb#L677-L688) for directories, and more work will be required to thread that to the metadata class, because the terminus helper's [plural path2instances method doesn't forward any options](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_serving/terminus_helper.rb#L27-L29).

One possibility is to have the `:file_server` terminus do the check after it's [asked the mount to resolve the path](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/file_server.rb#L29). That way the inline check is only called on the server, and we don't even bother creating a metadata object (in the case we can't inline). However, the `:file_server` implementation is shared for both `metadata` and `content` models, which is terrible and is why `content` implements [`collect`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_serving/content.rb#L24-L27). We could do something hacky in `:file_server` like:

```ruby
def find(..)
  ...
  return nil unless path = mount.find(relative_path, request)
  return nil unless inlineable?(request, path)
  path2instance(request, path)
end

def inlineable?(request, path)
   if model == Puppet::FileServing::Metadata
    # check if path starts with request.environment dir
  else
    true
  end
end
```

That's pretty aweful too. Ideally `content` and `metadata` would have different `file_server` implementations. But those indirected types delegate to a single [`selector`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_serving/metadata.rb#L13). So we'd need to provide different selectors, e.g. `indirects :file_metadata, :terminus_class => :metadata_selector`, and then have `metadata_selector` delegate to `metadata_file_server` for puppet URIs. That seems like it would work, but seems like a lot of trouble.

A third option is not worry about the extra file stat, and just have the compiler do:

```
  metadata = file.parameter(:source).metadata
  if metadata && metadata.path.start_with?(env_path)
    replace_metadata(resource, metadata)
  end
```

About PUP-5814, the compiler has access to `resource[:links]` and `resource[:target]` (in the `manage` case).

The downside is we'd be performing an extra stat for any file resource whose source parameter was a puppet URI and resolved to outside the per-env code dir. I'm guessing that's a small number of files as compared to the number of files sourced from per-env modules?